server/src/main/java/org/apache/iotdb/db/metadata/mnode/iterator/AbstractTraverserIterator.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iotdb.db.metadata.mnode.iterator;
+
+import org.apache.iotdb.commons.exception.MetadataException;
+import org.apache.iotdb.db.metadata.mnode.IMNode;
+import org.apache.iotdb.db.metadata.mtree.store.IMTreeStore;
+import org.apache.iotdb.db.metadata.template.Template;
+
+import java.util.Iterator;
+import java.util.Map;
+import java.util.NoSuchElementException;
+
+import static org.apache.iotdb.db.metadata.MetadataConstant.NON_TEMPLATE;
+
+/**
+ * ChildrenIterator for traverser. The iterator returns the filtered child nodes in the mtree and
+ * the child nodes in the template.
+ */
+public abstract class AbstractTraverserIterator implements IMNodeIterator {
+  private final IMNodeIterator directChildrenIterator;
+  private Iterator<IMNode> templateChildrenIterator;
+  protected IMNode nextMatchedNode;
+  protected boolean usingDirectChildrenIterator = true;
+  // if true, the pre deleted measurement or pre deactivated template won't be processed
+  private boolean skipPreDeletedSchema = false;
```

Comment:
It seems the ```templateChildrenIterator```, ```usingDirectChildrenIterator``` and ```skipPreDeletedSchema``` are all parameters used for measurement traverse when the parent is an entity. Should we encapsulate these into a subclass like ```MeasurementTraverserIterator``` ?

Reply:
done